File: frontend/src/types/deploy.ts
export type DeployStatus = 'pending' | 'success' | 'failed' | 'rolled_back'

export interface DeployRequest {
  router_ids: string[]
  template_id: string | null
  rendered_config: string
}

export interface DeployJobResponse {
  job_id: string
  history_ids: string[]
}

export interface RollbackJobResponse {
  job_id: string
  history_id: string
}

export interface HistoryListItem {
  id: string
  router_id: string
  router_hostname: string | null
  router_ip: string | null
  template_id: string | null
  deployed_by: string | null
  status: DeployStatus
  job_id: string | null
  deployed_at: string
  can_rollback: boolean
}

export interface HistoryDetail extends HistoryListItem {
  rendered_config: string | null
  config_snapshot: string | null
  output: string | null
}

export interface HistoryListResponse {
  items: HistoryListItem[]
  total: number
}

export interface BulkDeployResult {
  history_id: string
  router_id: string
  hostname?: string
  status: DeployStatus
  output?: string
}

export interface BulkDeployTaskResult {
  results: BulkDeployResult[]
}

File: backend/app/schemas/deploy.py
import uuid
from datetime import datetime
from pydantic import BaseModel, field_validator
from app.models.config_history import DeployStatus


class DeployRequest(BaseModel):
    router_ids: list[uuid.UUID]
    template_id: uuid.UUID | None = None
    rendered_config: str

    @field_validator("router_ids")
    @classmethod
    def at_least_one(cls, v: list) -> list:
        if not v:
            raise ValueError("At least one router must be specified")
        return v

    @field_validator("rendered_config")
    @classmethod
    def config_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("rendered_config cannot be empty")
        return v


class DeployJobResponse(BaseModel):
    job_id: str
    history_ids: list[uuid.UUID]


class RollbackJobResponse(BaseModel):
    job_id: str
    history_id: uuid.UUID


class HistoryListItem(BaseModel):
    id: uuid.UUID
    router_id: uuid.UUID
    router_hostname: str | None
    router_ip: str | None
    template_id: uuid.UUID | None
    deployed_by: uuid.UUID | None
    status: DeployStatus
    job_id: str | None
    deployed_at: datetime
    can_rollback: bool


class HistoryDetail(HistoryListItem):
    rendered_config: str | None
    config_snapshot: str | None
    output: str | None


class HistoryListResponse(BaseModel):
    items: list[HistoryListItem]
    total: int
